src/AzSK.ADO/Framework/Helpers/RateLimitHelper.ps1

```diff
@@ -0,0 +1,185 @@
+Set-StrictMode -Version Latest 
+class RateLimitHelper {
+
+    static $RateLimitEntity = [PSCustomObject]@{
+        RetryDate = $null
+        RateLimitRetryAfter = $null
+        RateLimitDelay = $null
+        RateLimitReset = $null
+    }
+    static $APIThrottleCounter = 0
+    static $MaxAllowedDelay = $null
+    static $MaxAPIThrottledBeforeSleep = $null
+    
+    <#
+        Method to retrieve headers related to rate limiter, store them in rate limit entity variable and send this to telemetry
+        Will work only if either UseRateLimit is set to true (in which case it will store appropriate headers in rate limit entity to be used for rate limiting subsequent API calls and also send this data to telemetry) or
+        RecordRateLimitDelay is set to true (in which case headers are stored only for telemetry purposes)    
+    #>
+    static UpdateRateLimitEntity($requestResult,$url){
+        try{
+            if($env:UseRateLimit -eq $true -or $env:RecordRateLimitDelay -eq $true){
+                $responseHeaders = $requestResult.Headers
+                $retryAfter = [RateLimitHelper]::GetHeaderValue($responseHeaders, "Retry-After");        
+                $rateLimitReset = [RateLimitHelper]::GetHeaderValue($responseHeaders, "X-RateLimit-Reset");
+                $rateLimitRemaining = [RateLimitHelper]::GetHeaderValue($responseHeaders, "X-RateLimit-Remaining");
+                $rateLimitDelay = [RateLimitHelper]::GetHeaderValue($responseHeaders, "X-RateLimit-Delay");
+                <#
+                    In case retry after is not null, API has been throttled, get all other headers for rate limiting
+                    In case retry after is null, there can be two conditions:
+                    1. API has not been throttled (i.e. rate limit remaining is null or greater than 0)
+                    2. API has been throttled (i.e. rate limit remaining is 0 and rate limit reset is the time when usage will be reset)
+                #> 
+                if([string]::IsNullOrEmpty($retryAfter)){
+                    if([string]::IsNullOrEmpty($rateLimitRemaining) -or ($rateLimitRemaining -ne 0 -or [string]::IsNullOrEmpty($rateLimitReset))){
+                        [RateLimitHelper]::APIThrottleCounter = 0;
+                        return;
+                    }            
+                } 
+                #Default behaviour is to ignore all delays less than 1 sec to improve scan time, can be overrrided by OverrideRateLimitThreshold
+                if($env:OverrideRateLimitThreshold -ne $true){
+                    if(-not [string]::IsNullOrEmpty($rateLimitDelay) -and $rateLimitDelay -lt 1){
+                        [RateLimitHelper]::APIThrottleCounter = 0;
+                        return;
+                    }
+                }                  
+                $rateLimitResetDate = (Get-Date 01.01.1970) + ([System.TimeSpan]::FromSeconds($rateLimitReset));
```

Comment:
Why this specific date?

Reply:
The rate limit reset from API is in the UNIX epoch format. To convert it to current UTC time, need to add this value to the UNIX epoch start date.

Reply:
Okay, please add this add comment in the code as well